syft/pkg/cataloger/java/archive_parser.go

```diff
@@ -304,40 +304,65 @@ func artifactIDMatchesFilename(artifactID, fileName string) bool {
 	return strings.HasPrefix(artifactID, fileName) || strings.HasSuffix(fileName, artifactID)
 }
 
-func findPomLicenses(pomProjectObject *parsedPomProject) {
+func findPomLicenses(pomProjectObject *parsedPomProject, cfg Config) {
 	// If we don't have any licenses until now, and if we have a parent Pom, then we'll check the parent pom in maven central for licenses.
 	if pomProjectObject != nil && pomProjectObject.Parent != nil && len(pomProjectObject.Licenses) == 0 {
-		parentPom, err := getPomFromMavenCentral(pomProjectObject.Parent.GroupID, pomProjectObject.Parent.ArtifactID, pomProjectObject.Parent.Version)
+		parentLicenses, err := recursivelyFindLicensesFromParentPom(
+			pomProjectObject.Parent.GroupID,
+			pomProjectObject.Parent.ArtifactID,
+			pomProjectObject.Parent.Version,
+			cfg)
 		if err != nil {
 			// We don't want to abort here as the parent pom might not exist in Maven Central, we'll just log the error
 			log.Tracef("unable to get parent pom from Maven central: %v", err)
 			return
 		}
-		parentLicenses := parseLicensesFromPom(parentPom)
-		if len(parentLicenses) > 0 || parentPom == nil || parentPom.Parent == nil {
+		if len(parentLicenses) > 0 {
 			for _, licenseName := range parentLicenses {
 				pomProjectObject.Licenses = append(pomProjectObject.Licenses, pkg.NewLicenseFromFields(licenseName, "", nil))
 			}
 		}
 	}
 }
 
-func formatMavenPomURL(groupID, artifactID, version string) (requestURL string, err error) {
+func formatMavenPomURL(groupID, artifactID, version, mavenBaseURL string) (requestURL string, err error) {
 	// groupID needs to go from maven.org -> maven/org
 	urlPath := strings.Split(groupID, ".")
 	artifactPom := fmt.Sprintf("%s-%s.pom", artifactID, version)
 	urlPath = append(urlPath, artifactID, version, artifactPom)
 
 	// ex:"https://repo1.maven.org/maven2/groupID/artifactID/artifactPom
-	requestURL, err = url.JoinPath(MavenBaseURL, urlPath...)
+	requestURL, err = url.JoinPath(mavenBaseURL, urlPath...)
 	if err != nil {
 		return requestURL, fmt.Errorf("could not construct maven url: %w", err)
 	}
 	return requestURL, err
 }
 
-func getPomFromMavenCentral(groupID, artifactID, version string) (*gopom.Project, error) {
-	requestURL, err := formatMavenPomURL(groupID, artifactID, version)
+func recursivelyFindLicensesFromParentPom(groupID, artifactID, version string, cfg Config) ([]string, error) {
+	var licenses []string
+	// As there can be nested parent poms, we'll recursively check for licenses until we reach the max depth
+	for i := 0; i < cfg.MaxParentRecursiveDepth; i++ {
```

Comment:
Instead of having a max depth to fetch parents, can we just check for cycles instead (probably just a simple string check of `groupId + ":" + artifactId` would work for this)? (Having a max parent depth also isn't a bad thing to do, but I think it could be larger if we are making sure there aren't cycles.

Reply:
I like the authors choice of having a max depth. I don't think anything we write at this point in the PR is going to guarantee to detect cycles perfectly, and would rather have a hard cutoff where it's guaranteed to short circuit.

Reply:
https://github.com/anchore/syft/issues/2284